File: platform/vcs-impl/src/com/intellij/vcs/commit/SaveCommittingDocumentsVetoer.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.vcs.commit;

import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.fileEditor.FileDocumentManagerListener;
import com.intellij.openapi.fileEditor.FileDocumentSynchronizationVetoer;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.util.UserDataHolderEx;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

final class SaveCommittingDocumentsVetoer extends FileDocumentSynchronizationVetoer implements FileDocumentManagerListener {

  private static final Object SAVE_DENIED = new Object();

  @Override
  public void beforeAllDocumentsSaving() {
    Map<Document, Project> documentsToWarn = getDocumentsBeingCommitted();
    if (!documentsToWarn.isEmpty()) {
      boolean allowSave = showAllowSaveDialog(documentsToWarn);
      updateSaveability(documentsToWarn, allowSave);
    }
  }

  @Override
  public boolean maySaveDocument(@NotNull Document document, boolean isSaveExplicit) {
    final Object beingCommitted = document.getUserData(AbstractCommitter.DOCUMENT_BEING_COMMITTED_KEY);
    if (beingCommitted == SAVE_DENIED) {
      return false;
    }
    if (beingCommitted instanceof Project) {
      return showAllowSaveDialog(Collections.singletonMap(document, (Project)beingCommitted));
    }
    return true;
  }

  private static Map<Document, Project> getDocumentsBeingCommitted() {
    Map<Document, Project> documentsToWarn = new HashMap<>();
    for (Document unsavedDocument : FileDocumentManager.getInstance().getUnsavedDocuments()) {
      final Object data = unsavedDocument.getUserData(AbstractCommitter.DOCUMENT_BEING_COMMITTED_KEY);
      if (data instanceof Project) {
        documentsToWarn.put(unsavedDocument, (Project)data);
      }
    }
    return documentsToWarn;
  }

  private static void updateSaveability(Map<Document, Project> documentsToWarn, boolean allowSave) {
    Object newValue = allowSave ? null : SAVE_DENIED;
    for (Document document : documentsToWarn.keySet()) {
      Project oldData = documentsToWarn.get(document);
      //the committing thread could have finished already and file is not being committed anymore
      ((UserDataHolderEx)document).replace(AbstractCommitter.DOCUMENT_BEING_COMMITTED_KEY, oldData, newValue);
    }
  }

  private static boolean showAllowSaveDialog(Map<Document, Project> documentsToWarn) {
    StringBuilder messageBuilder = new StringBuilder("The following " + (documentsToWarn.size() == 1 ? "file is" : "files are") +
                                                     " currently being committed to the VCS. " +
                                                     "Saving now could cause inconsistent data to be committed.\n");
    for (Document document : documentsToWarn.keySet()) {
      final VirtualFile file = FileDocumentManager.getInstance().getFile(document);
      messageBuilder.append(FileUtil.toSystemDependentName(file.getPath())).append("\n");
    }
    messageBuilder.append("Save the ").append(documentsToWarn.size() == 1 ? "file" : "files").append(" now?");

    Project project = documentsToWarn.values().iterator().next();
    int rc = Messages.showOkCancelDialog(project, messageBuilder.toString(), "Save Files During Commit", "Save Now", "Postpone Save",
                                         Messages.getQuestionIcon());
    return rc == Messages.OK;
  }
}

File: platform/vcs-impl/src/com/intellij/vcs/commit/SaveCommittingDocumentsVetoer.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.vcs.commit

import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileDocumentManagerListener
import com.intellij.openapi.fileEditor.FileDocumentSynchronizationVetoer
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.ui.Messages.getQuestionIcon
import com.intellij.openapi.ui.Messages.showOkCancelDialog
import com.intellij.openapi.util.Key
import com.intellij.openapi.util.UserDataHolderEx
import com.intellij.openapi.vcs.FilePath
import com.intellij.openapi.vcs.VcsBundle.message
import com.intellij.openapi.vcs.changes.Change
import com.intellij.openapi.vcs.changes.ChangesUtil
import com.intellij.openapi.vfs.VirtualFile

private sealed class SaveState
private object SaveDenied : SaveState()
private class RequireConfirmSave(val project: Project) : SaveState()

private val SAVE_STATE_KEY = Key<SaveState>("Vcs.Commit.SaveState")

private fun getSaveState(file: VirtualFile): SaveState? = file.getUserData(SAVE_STATE_KEY)
private fun setSaveState(file: VirtualFile, state: SaveState?) = file.putUserData(SAVE_STATE_KEY, state)
private fun replaceSaveState(file: VirtualFile, expectedOldState: SaveState?, newState: SaveState?) =
  (file as UserDataHolderEx).replace(SAVE_STATE_KEY, expectedOldState, newState)

internal class SaveCommittingDocumentsVetoer : FileDocumentSynchronizationVetoer(), FileDocumentManagerListener {
  override fun beforeAllDocumentsSaving() {
    val fileDocumentManager = FileDocumentManager.getInstance()
    val unsavedFiles = fileDocumentManager.unsavedDocuments
      .mapNotNull { document -> fileDocumentManager.getFile(document) }

    val protectedFiles = buildMap {
      for (file in unsavedFiles) {
        val state = getSaveState(file) as? RequireConfirmSave ?: continue
        put(file, state)
      }
    }
    if (protectedFiles.isEmpty()) return

    val project = protectedFiles.values.first().project
    val saveConfirmed = confirmSave(project, protectedFiles.keys)

    // use `replace` as commit could already be finished
    val newState = if (saveConfirmed) null else SaveDenied
    for ((file, oldState) in protectedFiles) {
      replaceSaveState(file, oldState, newState)
    }
  }

  override fun maySaveDocument(document: Document, isSaveExplicit: Boolean): Boolean {
    val file = FileDocumentManager.getInstance().getFile(document) ?: return true
    val saveState = getSaveState(file) ?: return true
    when (saveState) {
      SaveDenied -> return false
      is RequireConfirmSave -> return confirmSave(saveState.project, listOf(file))
    }
  }
}

fun vetoDocumentSaving(project: Project, changes: Collection<Change>, block: () -> Unit) {
  vetoDocumentSavingForPaths(project, ChangesUtil.getPaths(changes), block)
}

fun vetoDocumentSavingForPaths(project: Project, filePaths: Collection<FilePath>, block: () -> Unit) {
  val files = filePaths.mapNotNull { it.virtualFile }

  val confirmSaveState = RequireConfirmSave(project)
  files.forEach { setSaveState(it, confirmSaveState) }
  try {
    block()
  }
  finally {
    files.forEach { setSaveState(it, null) }
  }
}

private fun confirmSave(project: Project, files: Collection<VirtualFile>): Boolean {
  val text = message("save.committing.files.confirmation.text", files.size, files.joinToString("\n") { it.presentableUrl })

  return Messages.OK == showOkCancelDialog(
    project,
    text,
    message("save.committing.files.confirmation.title"),
    message("save.committing.files.confirmation.ok"),
    message("save.committing.files.confirmation.cancel"),
    getQuestionIcon()
  )
}